### PR TITLE
add bndtools workspace template fragment

### PR DESCRIPTION
**Pull request by chrisrueger**
This adds a Workspace Template Fragment which can be used by bndtools users to create a new Bnd Workspace.

See https://bnd.bndtools.org/chapters/620-template-fragments.html for  more information.

## Next steps

After this folder is available on github (e.g. in master branch):

- we add an entry in https://github.com/bndtools/workspace-templates/blob/master/index.bnd pointing to this folder in the tycho repo

Onces that is done, all bnd tools users will see a new entry for Eclipse Tycho Maven in bndtools, like this:

image

# Why

It is just a useful addition to promote the ability of Tycho to build Bnd Workspaces.

### CODE REVIEW

File: bndtools-templates/.mvn/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+</project>
+```
+
+This is useful if you want to add other maven plugins to your build e.g. for code analysis or reporting.
+
+## Creating executable jar based on .bndrun files
+
+Tycho can also create an executable .jar file of your application based on the `.bndrun` file containing `-runbundles`, which you can then deploy and start.
+
+Just add the following line to you `.mvn/maven.config` file:
```
Comment (laeubi) [inline bndtools-templates/.mvn/README.md:46]:
```suggestion
Just add the following line to you `.mvn/maven.config` file or pass it on the maven commandline
```

File: bndtools-templates/.mvn/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+cd mybndworkspace
+mvn clean install
+```
+
+This should be all you need to build your bnd workspace with Maven / Tycho.
+
+It will automatically consider all `bnd.bnd` files.
+
+
+## Optional parent pom.xml
```
Comment (laeubi) [inline bndtools-templates/.mvn/README.md:22]:
```suggestion
## Optional configurator `pom.xml`
```

Even though this is technically a parent pom, in Tycho we usually reference it as a "configurator pom" as its main purpose is configuring the build.

File: bndtools-templates/.mvn/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	<groupId>com.example.mygroup</groupId>
+	<artifactId>parent</artifactId>
+	<version>1.0.0-SNAPSHOT</version>
+	<packaging>pom</packaging>
+</project>
+```
+
+This is useful if you want to add other maven plugins to your build e.g. for code analysis or reporting.
+
+## Creating executable jar based on .bndrun files
```
Comment (laeubi) [inline bndtools-templates/.mvn/README.md:42]: Maybe this part is better explained at https://github.com/bndtools/workspace-templates/blob/master/index.bnd

File: bndtools-templates/.mvn/README.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+See documentation for the [Tycho BND Plugin](https://tycho.eclipseprojects.io/doc/main/BndBuild.html) for more information. 
+
+## Building your workspace
+
+```
+cd mybndworkspace
+mvn clean install
+```
+
+This should be all you need to build your bnd workspace with Maven / Tycho.
```
Comment (laeubi) [inline bndtools-templates/.mvn/README.md:17]:
```suggestion
This id all you need to build your bnd workspace with Maven / Tycho.
```

File: bndtools-templates/.mvn/README.md (hunk 3 above)
Comment (chrisrueger) [reply bndtools-templates/.mvn/README.md]: Done

File: bndtools-templates/.mvn/README.md (hunk 4 above)
Comment (chrisrueger) [reply bndtools-templates/.mvn/README.md]: done

File: bndtools-templates/.mvn/README.md (hunk 1 above)
Comment (chrisrueger) [reply bndtools-templates/.mvn/README.md]: done

**Comment by chrisrueger** [comment]
All suggestions fixed. I moved more stuff to BndBuild.md

File: setup/bnd-templates/pomless/.mvn/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# Tycho Workspace template fragment for bndtools for pomless Maven build
+
+The is a [bndtools template fragment](https://bnd.bndtools.org/chapters/620-template-fragments.html) which you can use in a new or existing bndtools workspace,
```
Comment (laeubi) [inline setup/bnd-templates/pomless/.mvn/README.md:3]:
```suggestion
This is a [bndtools template fragment](https://bnd.bndtools.org/chapters/620-template-fragments.html) which you can use in a new or existing bndtools workspace, 
```

File: setup/bnd-templates/pomless/.mvn/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
+# Tycho Workspace template fragment for bndtools for pomless Maven build
+
+The is a [bndtools template fragment](https://bnd.bndtools.org/chapters/620-template-fragments.html) which you can use in a new or existing bndtools workspace, 
+to add a Maven build based on Eclipse Tycho for your bnd workspace.
+
+All it does is adding a `.mvn` folder to your bnd workspace, contain a `extensions.xml` and a `maven.config`.
```
Comment (laeubi) [inline setup/bnd-templates/pomless/.mvn/README.md:6]:
```suggestion
It creates a `.mvn` folder in your bnd workspace root, contain a `extensions.xml` and a `maven.config`.
```

**Review by laeubi — COMMENTED** [review]
It overall looks good to me, please squash your commits into one so we can easier back-port this!

File: src/site/markdown/BndBuild.md [hunk 7]
```diff
@@ -51,6 +52,55 @@ If you want to further configure the build can be done in these ways:
...
+
+
 ## Mixed Builds
 
 You can even combine a BND Workspace and PDE bundles in a build, see [demo](https://github.com/eclipse-tycho/tycho/tree/master/demo/bnd-pde-workspace).
+
+
+## Troubleshooting
+
+Note: All properties in `.mvn/maven.config` can also be supplied via command line, for example to use a different tycho version use:
```
Comment (merks) [inline src/site/markdown/BndBuild.md:87]:
```suggestion
Note: All properties in `.mvn/maven.config` can also be supplied via command line, for example to use a different tycho version use the following command line argument:
```

File: src/site/markdown/BndBuild.md [hunk 8]
```diff
@@ -51,6 +52,55 @@ If you want to further configure the build can be done in these ways:
...
+
+`mvn clean install -Dtycho-version=5.0.0-SNAPSHOT`
+
+This uses the snapshot build which is e.g. useful if you build tycho yourself on your local machine and want to test that build.
```
Comment (merks) [inline src/site/markdown/BndBuild.md:91]:
```suggestion
This uses the snapshot build which is useful if you build tycho yourself on your local machine and want to test that build.
```

File: src/site/markdown/BndBuild.md [hunk 9]
```diff
@@ -51,6 +52,55 @@ If you want to further configure the build can be done in these ways:
...
+
+
+### polyglot.dump.pom
+
+E.g. `-Dpolyglot.dump.pom=pom-gen.xml` specifies that the generated pom.xml files which tycho polyglot creates is called `pom-gen.xml`.
```
Comment (merks) [inline src/site/markdown/BndBuild.md:96]:
```suggestion
E.g., `-Dpolyglot.dump.pom=pom-gen.xml` specifies that the generated pom.xml files which tycho polyglot creates is called `pom-gen.xml`. 
```

File: src/site/markdown/BndBuild.md [hunk 10]
```diff
@@ -51,6 +52,55 @@ If you want to further configure the build can be done in these ways:
...
+
+### polyglot.dump.pom
+
+E.g. `-Dpolyglot.dump.pom=pom-gen.xml` specifies that the generated pom.xml files which tycho polyglot creates is called `pom-gen.xml`. 
+This might be needed if you have name clashes with existing files in your build.
+
+
+### tycho.pomless.aggregator.names
+
+`-Dtycho.pomless.aggregator.names=_dummydisabled_` is for handling a rare edge case:
```
Comment (merks) [inline src/site/markdown/BndBuild.md:102]:
```suggestion
The command line argument `-Dtycho.pomless.aggregator.names=_dummydisabled_` is for handling a rare edge case: 
```

File: src/site/markdown/BndBuild.md [hunk 11]
```diff
@@ -51,6 +52,55 @@ If you want to further configure the build can be done in these ways:
...
+### polyglot.dump.pom
+
+E.g. `-Dpolyglot.dump.pom=pom-gen.xml` specifies that the generated pom.xml files which tycho polyglot creates is called `pom-gen.xml`. 
+This might be needed if you have name clashes with existing files in your build.
+
+
+### tycho.pomless.aggregator.names
+
+`-Dtycho.pomless.aggregator.names=_dummydisabled_` is for handling a rare edge case: 
+In case you have your bnd workspace in a subfolder called one of `bundles`,`plugins`,`tests`,`features`,`sites`,`products`,`releng` then tycho can have problems.
```
Comment (merks) [inline src/site/markdown/BndBuild.md:103]:
```suggestion
In case you have your bnd workspace in a subfolder called named `bundles`, `plugins`, `tests`,`features`, `sites`,`products`, or `releng` then tycho can have problems.
```

File: src/site/markdown/BndBuild.md [hunk 12]
```diff
@@ -51,6 +52,55 @@ If you want to further configure the build can be done in these ways:
...
+The reason is that those names are kind of 'magic' names reserved for special usecases. 
+`-Dtycho.pomless.aggregator.names=_dummydisabled_` is a workaround to instruct Tycho to ignore those names. You can choose any word like `_dummydisabled_`, but just make sure you do not have a folder like this.
```
Comment (merks) [inline src/site/markdown/BndBuild.md:105]:
```suggestion
`-Dtycho.pomless.aggregator.names=_dummydisabled_` is a workaround to instruct Tycho to ignore those names. You can choose any word like `_dummydisabled_`, but just make sure you do not have a folder with that name.
```

**Review by merks — COMMENTED** [review]
I think the suggest changes will improve the documentation (of this cool new feature).

**Comment by chrisrueger** [comment]
Thanks a lot @merks for the readme suggestions. I committed them all.
@laeubi I squashed all commits. Ready to go I think.

**Comment by laeubi** [comment]
Build will not verify this ..
